components/external_storage/src/s3.rs

```diff
@@ -0,0 +1,190 @@
+    client: S3Client,
+}
+
+impl S3Storage {
+    /// Create a new S3 storage for the given config.
+    pub fn new(config: &Config) -> Result<S3Storage> {
+        // This can greatly improve performance dealing with payloads greater
+        // than 100MB. See https://github.com/rusoto/rusoto/pull/1227
+        // for more information.
+        let mut http_config = HttpConfig::new();
+        http_config.read_buf_size(1024 * 1024 * 2);
+        let http_dispatcher = HttpClient::new_with_config(http_config).unwrap();
+
+        S3Storage::with_request_dispatcher(config, http_dispatcher)
+    }
+
+    fn with_request_dispatcher<D>(config: &Config, dispatcher: D) -> Result<S3Storage>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+        D::Future: Send,
+    {
+        if config.bucket.is_empty() {
+            return Err(Error::new(ErrorKind::InvalidInput, "missing bucket name"));
+        }
+        let region = if config.endpoint.is_empty() {
+            config.region.parse::<region::Region>().map_err(|e| {
+                Error::new(
+                    ErrorKind::InvalidInput,
+                    format!("invalid region format {}: {}", config.region, e),
+                )
+            })?
+        } else {
+            region::Region::Custom {
+                name: config.region.clone(),
+                endpoint: config.endpoint.clone(),
+            }
+        };
+        let client = if config.access_key.is_empty() || config.secret_access_key.is_empty() {
+            let cred_provider = DefaultCredentialsProvider::new().map_err(|e| {
+                Error::new(
+                    ErrorKind::PermissionDenied,
+                    format!("unable to get credentials: {}", e),
+                )
+            })?;
+            S3Client::new_with(dispatcher, cred_provider, region)
+        } else {
+            let cred_provider = StaticProvider::new(
+                config.access_key.clone(),
+                config.secret_access_key.clone(),
+                None, /* token */
+                None, /* valid_for */
+            );
+            S3Client::new_with(dispatcher, cred_provider, region)
+        };
+        Ok(S3Storage {
+            config: config.clone(),
+            client,
+        })
+    }
+
+    fn maybe_prefix_key(&self, key: &str) -> String {
+        if !self.config.prefix.is_empty() {
+            return format!("{}/{}", self.config.prefix, key);
+        }
+        key.to_owned()
+    }
+}
+
+impl ExternalStorage for S3Storage {
+    fn write(&self, name: &str, reader: &mut dyn Read) -> Result<()> {
+        let mut content = vec![];
+        reader.read_to_end(&mut content)?;
+        let key = self.maybe_prefix_key(name);
+        debug!("save file to s3 storage"; "key" => %key);
+        let req = PutObjectRequest {
+            key,
+            bucket: self.config.bucket.clone(),
+            body: Some(content.into()),
+            acl: Some(self.config.acl.clone()),
```

Comment:
I think all these string fields need to check empty. If the string is empty, you need to pass `None` or use default in the PutObjectRequest.